pgrx/src/spi/tuple.rs

```diff
@@ -8,19 +8,56 @@ use crate::memcxt::PgMemoryContexts;
 use crate::pg_sys::panic::ErrorReportable;
 use crate::pg_sys::{self, PgOid};
 use crate::prelude::*;
+use crate::spi::SpiClient;
 
-use super::{SpiError, SpiErrorCodes, SpiOkCodes, SpiResult};
+use super::{SpiError, SpiErrorCodes, SpiResult};
 
 #[derive(Debug)]
 pub struct SpiTupleTable<'client> {
-    #[allow(dead_code)]
-    pub(super) status_code: SpiOkCodes,
-    pub(super) table: Option<&'client mut pg_sys::SPITupleTable>,
-    pub(super) size: usize,
-    pub(super) current: isize,
+    // SpiTupleTable borrows global state setup by the active SpiClient.  It doesn't use the client
+    // directly, but we need to make sure we don't outlive it, so here it is
+    _client: PhantomData<&'client SpiClient>,
+
+    // and this is that global state.  In ::wrap(), this comes from whatever the current value of
+    // `pg_sys::SPI_tuptable` happens to be.  Postgres may change where SPI_tuptable points
+    // throughout the lifetime of an active SpiClient, but it doesn't mutate (or deallocate) what
+    // it happens to point to  This allows us to have multiple active SpiTupleTables
+    // within a Spi connection.  Whatever this points to is freed via `pg_sys::SPI_freetuptable()`
+    // when we're dropped.
+    table: Option<NonNull<pg_sys::SPITupleTable>>,
+    size: usize,
+    current: isize,
 }
 
 impl<'client> SpiTupleTable<'client> {
+    /// Wraps the current global `pg_sys::SPI_tuptable` as a new [`SpiTupleTable`] instance, with
+    /// a lifetime tied to the specified [`SpiClient`].
+    pub(super) fn wrap(_client: &'client SpiClient, last_spi_status_code: i32) -> SpiResult<Self> {
+        Spi::check_status(last_spi_status_code)?;
+
+        unsafe {
+            //
+            // SAFETY:  The unsafeness here is that we're accessing static globals.  Fortunately,
+            // Postgres is not multi-threaded so we're okay to do this
+            //
```

Comment:
*nervous laughter*

Reply:
Future Eric is gonna love this comment, if this lands.